tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+	chkData := sgmBytes.Range(chkDataEnd-int(chkDataLen), chkDataEnd)
+	return cdm.pool.Get(chunkenc.Encoding(chkEnc), chkData)
+}
+
+func (cdm *ChunkDiskMapper) getChunkFromBuffer(chunkRef uint64) chunkenc.Chunk {
+	shardIdx := chunkRef % inBufferShards
+
+	cdm.inBufferChunksLocks[shardIdx].RLock()
+	defer cdm.inBufferChunksLocks[shardIdx].RUnlock()
+
+	return cdm.inBufferChunks[shardIdx][chunkRef]
+}
+
+// IterateAllChunks iterates on all the chunks in its byte slices in the order of the head chunk file sequence
+// and runs the provided function on each chunk. It returns on the first error encountered.
+func (cdm *ChunkDiskMapper) IterateAllChunks(f func(seriesRef, chunkRef uint64, mint, maxt int64) error) error {
+	chkCRC32 := newCRC32()
+
+	// Iterate files in ascending order.
+	seqs := make([]int, 0, len(cdm.mmappedChunkFiles))
+	for seg := range cdm.mmappedChunkFiles {
+		seqs = append(seqs, seg)
+	}
+	sort.Ints(seqs)
+	for _, seq := range seqs {
+		bs := cdm.mmappedChunkFiles[seq]
+		sliceLen := bs.Len()
+		idx := HeadChunkFileHeaderSize
+		for idx < sliceLen {
+			if sliceLen-idx < 30 {
+				return &CorruptionErr{
+					Dir:       cdm.dir.Name(),
+					FileIndex: seq,
+					Err:       errors.Errorf("head chunk file doesn't include enough bytes to read the chunk header - required:%v, available:%v, file:%d", idx+35, sliceLen, seq),
+				}
+			}
+			chkCRC32.Reset()
+			startIdx := idx
+			seriesRef := binary.BigEndian.Uint64(bs.Range(idx, idx+8))
+			idx += 8
+
+			mint := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			maxt := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			chunkRef := uint64(seq)<<32 | uint64(idx)
+			if err := f(seriesRef, chunkRef, mint, maxt); err != nil {
+				return &CorruptionErr{
+					Dir:       cdm.dir.Name(),
+					FileIndex: seq,
+					Err:       err,
+				}
+			}
+
+			idx++ // Skip encoding.
+			// Skip the data.
+			dataLen, n := binary.Uvarint(bs.Range(idx, idx+MaxChunkLengthFieldSize))
+			idx += n + int(dataLen)
+
+			// Check CRC.
+			sum := bs.Range(idx, idx+CRCSize)
+			if _, err := chkCRC32.Write(bs.Range(startIdx, idx)); err != nil {
+				return err
+			}
+			if act := chkCRC32.Sum(nil); !bytes.Equal(act, sum) {
+				return errors.Errorf("checksum mismatch expected:%x, actual:%x", sum, act)
+			}
+
+			idx += CRCSize
+		}
+		if idx > sliceLen {
+			// It should be equal to the slice length.
+			return &CorruptionErr{
+				Dir:       cdm.dir.Name(),
+				FileIndex: seq,
+				Err:       errors.Errorf("head chunk file doesn't include enough bytes to read the last chunk data - required:%v, available:%v, file:%d", idx, sliceLen, seq),
+			}
+		}
+	}
+
+	return nil
+}
+
+// Truncate deletes the head chunk files which are strictly below the mint.
+// mint should be in milliseconds.
+func (cdm *ChunkDiskMapper) Truncate(mint int64) error {
+	var removedFiles []int
+
+	cdm.mmappedChunksMtx.RLock()
+	for seq, bs := range cdm.mmappedChunkFiles {
```

Comment:
When we have it sorted we can just abort when we see greater mint

Reply:
It's a map, so it won't be sorted always I guess :). Regarding using a slice instead, I have comment above it above.